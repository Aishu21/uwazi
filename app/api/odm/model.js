import mongoose from 'mongoose';
import { model as updateLogModel } from 'api/updatelog';
import models from './models';

const generateID = mongoose.Types.ObjectId;
export { generateID };

export default (collectionName, schema) => {
  const getAffectedIds = async conditions => mongoose.models[collectionName].find(conditions, { _id: true });

  const upsertLogOne = async (doc, next) => {
    const logData = { namespace: collectionName, mongoId: doc._id };
    await updateLogModel.findOneAndUpdate(logData, { ...logData, timestamp: Date.now() }, { upsert: true, lean: true });
    next();
  };

  async function upsertLogMany(affectedIds, deleted = false) {
    await updateLogModel.updateMany(
      { mongoId: { $in: affectedIds.map(i => i._id) }, namespace: collectionName },
      { $set: { timestamp: Date.now(), deleted } },
      { upsert: true, lean: true }
    );
  }

  schema.post('save', upsertLogOne);
  schema.post('findOneAndUpdate', upsertLogOne);

  schema.post('updateMany', async function updateMany(doc, next) {
    const affectedIds = await getAffectedIds(this._conditions);
    await upsertLogMany(affectedIds);
    next();
  });

  const MongooseModel = mongoose.model(collectionName, schema);

  const odmModel = {
    db: MongooseModel,
    save: (data) => {
      if (Array.isArray(data)) {
        const promises = data.map((entry) => {
          if (entry._id) {
            return MongooseModel.findOneAndUpdate({ _id: entry._id }, entry, { new: true, lean: true });
          }
          return MongooseModel.create(entry).then(saved => saved.toObject());
        });
        return Promise.all(promises);
      }

      if (data._id) {
        return MongooseModel.findOneAndUpdate({ _id: data._id }, data, { new: true, lean: true });
      }
      return MongooseModel.create(data).then(saved => saved.toObject());
    },

    get: (query, select = '', pagination = {}) => MongooseModel.find(query, select, Object.assign({ lean: true }, pagination)),

    count: condition => MongooseModel.count(condition),

    getById: id => MongooseModel.findById(id, {}, { lean: true }),

    delete: async (condition) => {
      let cond = condition;
      if (mongoose.Types.ObjectId.isValid(condition)) {
        cond = { _id: condition };
      }
      const affectedIds = await getAffectedIds(cond);
      const result = await MongooseModel.deleteMany(cond);
      await upsertLogMany(affectedIds, true);

      return result;
    }
  };

  models[collectionName] = odmModel;
  return odmModel;
};
