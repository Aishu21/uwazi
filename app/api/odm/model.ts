import mongoose from "mongoose";

import { model as updatelogsModel } from "api/updatelogs";

import models from "./models";
import { OdmModel } from "./models";

const generateID = mongoose.Types.ObjectId;
export { generateID };

export default <T extends mongoose.Document>(
  collectionName: string,
  schema: mongoose.Schema
) => {
  const getAffectedIds = async (conditions: any) =>
    mongoose.models[collectionName].distinct("_id", conditions);

  const upsertLogOne = async (
    doc: mongoose.Document,
    next: (err?: mongoose.NativeError) => void
  ) => {
    const logData = { namespace: collectionName, mongoId: doc._id };
    await updatelogsModel.findOneAndUpdate(
      logData,
      { ...logData, timestamp: Date.now(), deleted: false },
      { upsert: true }
    );
    next();
  };

  async function upsertLogMany(affectedIds: any[], deleted = false) {
    await updatelogsModel.updateMany(
      { mongoId: { $in: affectedIds }, namespace: collectionName },
      { $set: { timestamp: Date.now(), deleted } },
      { upsert: true, lean: true }
    );
  }

  schema.post("save", upsertLogOne);
  schema.post("findOneAndUpdate", upsertLogOne);

  const MongooseModel = mongoose.model<T>(collectionName, schema);

  const saveOne = async (data: mongoose.Document): Promise<T | null> => {
    const documentExists = await MongooseModel.findById(data._id, "_id");

    if (documentExists) {
      return MongooseModel.findOneAndUpdate({ _id: data._id }, data, {
        new: true
      }).exec();
    }
    return MongooseModel.create(data).then(saved => saved.toObject());
  };

  const odmModel: OdmModel<mongoose.Document> = {
    db: MongooseModel,
    save: (data: mongoose.Document | mongoose.Document[]) => {
      if (Array.isArray(data)) {
        const promises = data.map(entry => saveOne(entry));
        return Promise.all(promises);
      }

      return saveOne(data);
    },

    get: (query: any, select = "", pagination = {}) =>
      MongooseModel.find(
        query,
        select,
        Object.assign({ lean: true }, pagination)
      ).exec(),

    count: (condition: any) => MongooseModel.count(condition).exec(),

    getById: (id: any | string | number) =>
      MongooseModel.findById(id, {}, { lean: true }).exec(),

    delete: async condition => {
      let cond = condition;
      if (mongoose.Types.ObjectId.isValid(condition)) {
        cond = { _id: condition };
      }

      const affectedIds = await getAffectedIds(cond);
      const result = await MongooseModel.deleteMany(cond);

      if (affectedIds.length) {
        await upsertLogMany(affectedIds, true);
      }

      return result;
    }
  };

  models[collectionName] = odmModel;
  return odmModel;
};
