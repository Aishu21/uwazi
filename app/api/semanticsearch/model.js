import mongoose from 'mongoose';
import instanceModel from 'api/odm';

const resultSentenceSchema = new mongoose.Schema({
  sentence: String,
  score: Number
});

const resultSchema = new mongoose.Schema({
  sharedId: String,
  status: String,
  results: [resultSentenceSchema]
});

const searchSchema = new mongoose.Schema({
  language: String,
  searchTerm: String,
  status: String,
  documents: [resultSchema]
});

let Model = mongoose.model('semanticsearches', searchSchema);
export default instanceModel(Model);
