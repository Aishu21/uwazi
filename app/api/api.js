import bodyParser from 'body-parser'
import db_config from './config/database.js'

export default app => {

  //set db to use
  db_config.db_url = db_config.production;

  //common middlewares
  app.use(bodyParser.json());

  //module routes
  require('./auth/routes.js')(app);

}
