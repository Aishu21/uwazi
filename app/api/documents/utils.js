export function updateMetadataNames(documents, nameMatches) {
  documents.forEach((doc) => {
    let oldMetadata = Object.assign({}, doc.metadata);
    Object.keys(doc.metadata).forEach((propertyName) => {
      if (nameMatches[propertyName]) {
        delete doc.metadata[propertyName];
        doc.metadata[nameMatches[propertyName]] = oldMetadata[propertyName];
      }
    });
  });

  return documents;
}
