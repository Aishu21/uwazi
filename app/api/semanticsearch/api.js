import fetch from 'isomorphic-fetch';
import SEMANTIC_SEARCH_URL from 'api/config/semanticsearch';

const semanticSearchAPI = {
  async processDocument(args) {
    const res = await fetch(`${SEMANTIC_SEARCH_URL}/semanticSearch/searchOneDoc`, {
      method: 'POST',
      body: JSON.stringify(args),
      headers: {
        'Content-Type': 'application/json'
      }
    });
    const body = await res.json();
    return body;
  }
};

export default semanticSearchAPI;
