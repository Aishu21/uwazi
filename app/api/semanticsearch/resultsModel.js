import mongoose from 'mongoose';
import instanceModel from 'api/odm';

const resultSchema = new mongoose.Schema({
  sharedId: String,
  searchId: String,
  status: String,
  results: [{
      sentence: String,
      score: Number,
      page: Number
  }]
});

const Model = mongoose.model('semanticsearchresults', resultSchema);
export default instanceModel(Model);
