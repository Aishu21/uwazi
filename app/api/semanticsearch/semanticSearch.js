import { promisify } from 'util';
import async from 'async';
import { Types } from 'mongoose';
import { search } from '../search';
import model from './model';
import resultsModel from './resultsModel';
import api from './api';
import documentsModel from '../documents';
import workers from './workerManager';
import { createError } from '../utils';

import date from 'api/utils/date.js';

export const PENDING = 'pending';
export const COMPLETED = 'completed';
export const PROCESSING = 'processing';
export const IN_PROGRESS = 'inProgress';
export const STOPPED = 'stopped';

const SEARCH_BATCH_SIZE = 5;

const eachLimitAsync = promisify(async.eachLimit);

const getSearchDocuments = async ({ documents, query }, language, user) => {
  if (documents && documents.length) {
    return documents;
  }
  const _query = { ...query, limit: 9999, searchTerm: '' };
  const res = await search.search(_query, language, user);
  return res.rows.map(document => document.sharedId);
};

const updateSearchDocumentStatus = async (searchId, sharedId, status) => model.db.findOneAndUpdate({
  _id: searchId,
  'documents.sharedId': sharedId
}, {
  $set: { 'documents.$.status': status }
}, { new: true, lean: true });

const setSearchDocumentResults = async (searchId, sharedId, results) => {
  const averageScore = results.reduce((total, curr) => total + curr.score, 0) / results.length;
  const docResults = await resultsModel.db.findOneAndUpdate({
    sharedId,
    searchId
  }, {
    sharedId,
    searchId,
    averageScore,
    results: results.sort((r1, r2) => r2.score - r1.score),
    status: COMPLETED
  }, { upsert: true, new: true });
  return docResults;
};

const removePageAnnotations = text => text.replace(/\[\[\d+\]\]/g, '');

const processDocument = async (searchId, searchTerm, sharedId, language) => {
  const [doc] = await documentsModel.get({ sharedId, language }, '+fullText');
  const { fullText } = doc;

  await updateSearchDocumentStatus(searchId, sharedId, PROCESSING);
  if (!fullText) {
    return updateSearchDocumentStatus(searchId, sharedId, COMPLETED);
  }
  Object.keys(fullText).forEach((page) => {
    fullText[page] = removePageAnnotations(fullText[page]);
  });
  const results = await api.processDocument({
    searchTerm,
    contents: fullText
  });
  const savedResults = await setSearchDocumentResults(searchId, sharedId, results);
  await updateSearchDocumentStatus(searchId, sharedId, COMPLETED);
  return savedResults;
};

const updateSearchStatus = (searchId, status) => model.save({
  _id: searchId,
  status
});

const getAllDocumentResults = async searchId => resultsModel.get({ searchId });

const processSearchLimit = async (searchId, docLimit) => {
  const searchObject = await updateSearchStatus(searchId, IN_PROGRESS);
  const { language, searchTerm } = searchObject;
  const docs = searchObject.documents
  .filter(doc => doc.status !== COMPLETED);
  const docsToSearch = docs.length > docLimit ?
    docs.slice(0, docLimit) : docs;
  await eachLimitAsync(docsToSearch, SEARCH_BATCH_SIZE, async doc => processDocument(
    searchId, searchTerm, doc.sharedId, language));
  const updatedSearch = await model.getById(searchId);
  const isNotDone = updatedSearch.documents.some(doc => doc.status !== COMPLETED);
  if (isNotDone) {
    return { updatedSearch, processedDocuments: docsToSearch.map(d => d.sharedId) };
  }
  return { updatedSearch: await updateSearchStatus(searchId, COMPLETED), processedDocuments: docsToSearch.map(d => d.sharedId) };
};

const create = async (args, language, user) => {
  const docs = await getSearchDocuments(args, language, user);
  const newSearch = {
    documents: docs.map(docId => ({
      sharedId: docId,
      status: PENDING
    })),
    status: PENDING,
    searchTerm: args.searchTerm,
    query: args.query,
    language,
    creationDate: date.currentUTC()
  };
  const savedSearch = await model.save(newSearch);
  workers.notifyNewSearch(savedSearch._id);
  return savedSearch;
};

const filterSearchResults = async (searchId, { skip = 0, limit = 50, threshold = 0.3, minRelevantSentences = 5 }) =>
  resultsModel.db.aggregate([
    {
      $match: { searchId: Types.ObjectId(searchId) }
    },
    {
      $project: {
        searchId: 1,
        sharedId: 1,
        status: 1,
        totalResults: { $size: '$results' },
        results: { $filter: { input: '$results', as: 'result', cond: { $gte: ['$$result.score', threshold] } } }
      }
    },
    {
      $project: {
        totalResults: 1,
        searchId: 1,
        sharedId: 1,
        status: 1,
        results: 1,
        numRelevant: { $size: '$results' },
        percentage: { $divide: [{ $size: '$results' }, '$totalResults'] }
      }
    },
    {
      $match: { numRelevant: { $gte: minRelevantSentences } }
    },
    {
      $sort: { percentage: -1 }
    },
    {
      $skip: skip
    },
    {
      $limit: limit
    }
  ]);

const getSearch = async (searchId) => {
  const theSearch = await model.getById(searchId);
  if (!theSearch) {
    throw createError('Search not found', 404);
  }
  const results = await filterSearchResults(searchId, {});
  const docIds = results.map(r => r.sharedId);
  const docs = await documentsModel.get({ sharedId: { $in: docIds }, language: theSearch.language });
  const docsWithResults = docs.map(doc => (
    {
      ...doc,
      semanticSearch: results.find(res => res.sharedId === doc.sharedId)
    }
  ));
  theSearch.results = docsWithResults;
  return theSearch;
};

const getDocumentResultsByIds = async (searchId, docIds) => {
  const theSearch = searchId._id ? searchId : await model.getById(searchId);
  const results = await resultsModel.get({ searchId, sharedId: { $in: docIds } });
  const docs = await documentsModel.get({ sharedId: { $in: docIds }, language: theSearch.language });
  const docsWithResults = docs.map(doc => (
    {
      ...doc,
      semanticSearch: results.find(res => res.sharedId === doc.sharedId)
    }
  ));
  return docsWithResults;
};

const getSearchesByDocument = async (docId) => {
  const results = await resultsModel.get({ sharedId: docId });
  const searchPromises = results.map(({ searchId }) => model.getById(searchId));
  const searches = await Promise.all(searchPromises);
  searches.map((theSearch, index) => ({
    ...theSearch,
    results: results[index]
  }));
  return searches;
};

const deleteSearch = async (searchId) => {
  const res = await model.delete(searchId);
  if (res.n !== 1) {
    throw createError('Search not found', 404);
  }
  await resultsModel.delete({ searchId });
  return { deleted: true };
};

const stopSearch = async (searchId) => {
  const res = await model.db.updateOne({
    _id: searchId,
    status: { $in: [IN_PROGRESS, PENDING] }
  }, {
    $set: { status: STOPPED }
  });
  if (!res.n) {
    throw createError('Only running searches can be stopped', 400);
  }
  return model.getById(searchId);
};

const resumeSearch = async (searchId) => {
  const res = await model.db.updateOne({
    _id: searchId,
    status: { $in: [STOPPED] }
  }, {
    $set: { status: PENDING }
  });
  if (!res.n) {
    throw createError('Only stopped searches can be resumed', 400);
  }
  workers.notifyNewSearch(searchId);
  return model.getById(searchId);
};

const getAllSearches = () => model.get().sort('-creationDate');
const getInProgress = async () => model.get({ status: IN_PROGRESS }).sort('-creationDate');
const getPending = async () => model.get({ status: PENDING }).sort('-creationDate');

const semanticSearch = {
  _id: Math.random(),
  create,
  processDocument,
  processSearchLimit,
  getAllDocumentResults,
  getDocumentResultsByIds,
  getAllSearches,
  getPending,
  getInProgress,
  getSearch,
  getSearchesByDocument,
  deleteSearch,
  stopSearch,
  resumeSearch
};

export default semanticSearch;
