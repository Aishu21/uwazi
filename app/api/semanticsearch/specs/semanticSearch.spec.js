import db from 'api/utils/testing_db';
import workers from '../workerManager';
import semanticSearch from '../semanticSearch';
import { search } from '../../search';
import model from '../model';
import resultsModel from '../resultsModel';
import api from '../api';

import fixtures from './fixtures';
import { search1Id, search2Id, search3Id, doc1Id, docWithoutTextId } from './fixtures';
import { createError } from '../../utils';

describe('semanticSearch', () => {
  beforeEach((done) => {
    db.clearAllAndLoad(fixtures).then(done);
  });
  afterAll((done) => {
    db.disconnect().then(done);
  });

  describe('create', () => {
    beforeEach(() => {
      jest.spyOn(workers, 'notifyNewSearch').mockImplementation(() => {});
    });
    it('should create and save new search', async () => {
      const args = { searchTerm: 'Test term', documents: ['doc1', 'doc2'] };
      const created = await semanticSearch.create(args, 'en', 'user');
      const savedSearch = await model.getById(created._id);
      delete savedSearch._id;
      expect(savedSearch).toMatchSnapshot();
    });
    it('should send the search to the workers', async () => {
      const args = { searchTerm: 'Test term', documents: ['doc1', 'doc2'] };
      const created = await semanticSearch.create(args, 'en', 'user');
      expect(workers.notifyNewSearch).toHaveBeenCalledWith(created._id);
    });
    describe('when query is provided instead of documents list', () => {
      it('should fetch the documents based on the query', async () => {
        jest.spyOn(search, 'search').mockResolvedValue({
          rows: [
            { sharedId: 'docA' },
            { sharedId: 'docB' }
          ]
        });
        const query = { filters: {} };
        const args = { searchTerm: 'term', query };
        const created = await semanticSearch.create(args, 'en', 'user');
        expect(search.search).toHaveBeenCalledWith(query, 'en', 'user');
        expect(created.documents).toEqual([
          { sharedId: 'docA', status: 'pending' },
          { sharedId: 'docB', status: 'pending' }
        ]);
      });
    });
  });

  describe('processDocument', () => {
    let expectedResults;
    beforeEach(() => {
      expectedResults = [
        { page: 1, text: 'page 1', score: 0.2 },
        { page: 2, text: 'page 2', score: 0.6 }
      ];
      jest.spyOn(api, 'processDocument').mockResolvedValue(expectedResults);
      api.processDocument.mockClear();
    });
    it('should send document to semantic search api for processing', async () => {
      await semanticSearch.processDocument(search1Id, 'legal', doc1Id, 'en');
      expect(api.processDocument).toHaveBeenCalledWith({
        searchTerm: 'legal',
        contents: {
          1: 'page 1',
          2: 'page 2'
        }
      });
    });
    it('should update the status of the document to be completed', async () => {
      await semanticSearch.processDocument(search1Id, 'legal', doc1Id, 'en');
      const theSearch = await model.getById(search1Id);
      const docInSearch = theSearch.documents.find(doc => doc.sharedId === doc1Id);
      expect(docInSearch.status).toBe('completed');
    });
    it('should save the results of the document sorted by score in descending order, and compute average score', async () => {
      await semanticSearch.processDocument(search1Id, 'legal', doc1Id, 'en');
      const [docResults] = await resultsModel.get({ searchId: search1Id, sharedId: doc1Id });
      expect(docResults.status).toBe('completed');
      expect(docResults.averageScore).toBe((0.2 + 0.6) / 2);
      expect(
        docResults.results
        .map(({ page, text, score }) => ({ page, text, score }))
      ).toMatchSnapshot();
    });
    describe('if document has no fullText', () => {
      it('should mark as completed without processing', async () => {
        await semanticSearch.processDocument(search1Id, 'legal', docWithoutTextId, 'en');
        expect(api.processDocument).not.toHaveBeenCalled();
        const theSearch = await model.getById(search1Id);
        const docInSearch = theSearch.documents.find(doc => doc.sharedId === docWithoutTextId);
        expect(docInSearch.status).toBe('completed');
        const docRes = await resultsModel.get({ searchId: search1Id, sharedId: docWithoutTextId });
        expect(docRes.length).toBe(0);
      });
    });
  });

  describe('processSearchLimit', () => {
    const expectedResults = [
      { page: 1, text: 'page 1', score: 0.6 },
      { page: 2, text: 'page 2', score: 0.2 }
    ];
    beforeEach(() => {
      jest.spyOn(api, 'processDocument').mockResolvedValue(expectedResults);
      api.processDocument.mockClear();
    });
    it('should process only up to specified number of unprocessed docs in the search', async () => {
      await semanticSearch.processSearchLimit(search2Id, 2);
      expect(api.processDocument).toHaveBeenCalledTimes(2);
      expect(api.processDocument).toHaveBeenCalledWith({
        searchTerm: 'injustice',
        contents: { 1: 'text2' }
      });
      expect(api.processDocument).toHaveBeenCalledWith({
        searchTerm: 'injustice',
        contents: { 1: 'text3' }
      });
      const theSearch = await model.getById(search2Id);
      expect(theSearch.documents.some(
        doc => doc.sharedId === 'doc2' && doc.status === 'completed')).toBe(true);
      expect(theSearch.documents.some(
        doc => doc.sharedId === 'doc3' && doc.status === 'completed')).toBe(true);
      expect(theSearch.documents.some(
        doc => doc.sharedId === 'doc5' && doc.status === 'pending')).toBe(true);
      const res = await resultsModel.get({ searchId: search2Id });
      expect(res.some(r => r.sharedId === 'doc2')).toBe(true);
      expect(res.some(r => r.sharedId === 'doc3')).toBe(true);

      expect(theSearch.status).toBe('inProgress');
    });
    it('should mark search as complete if all documents are processed', async () => {
      await semanticSearch.processSearchLimit(search2Id, 5);
      const theSearch = await model.getById(search2Id);
      expect(theSearch.status).toBe('completed');
    });
  });

  describe('getAllSearches', () => {
    it('should return all searches', async () => {
      const searches = await semanticSearch.getAllSearches();
      expect(searches[0]._id).toEqual(search1Id);
      expect(searches.length).toBe(3);
    });
  });

  describe('getPending', () => {
    it('should return pending searches', async () => {
      const pending = await semanticSearch.getPending();
      expect(pending.length).toBe(1);
      expect(pending[0]._id).toEqual(search2Id);
    });
  });

  describe('getInProgress', () => {
    it('should return all searches in progress', async () => {
      const inProgress = await semanticSearch.getInProgress();
      expect(inProgress.length).toBe(1);
      expect(inProgress[0]._id).toEqual(search1Id);
    });
  });

  describe('getAllDocumentResults', () => {
    it('should return all document results of the specified search', async () => {
      const results = await semanticSearch.getAllDocumentResults(search3Id);
      expect(results.length).toEqual(2);
      expect(results.some(doc => doc.sharedId === 'doc1')).toBe(true);
      expect(results.some(doc => doc.sharedId === 'doc2')).toBe(true);
    });
  });

  describe('getSearch', () => {
    it('should fetch a search by id and its document entities with semantic search results', async () => {
      const res = await semanticSearch.getSearch(search3Id);
      res.results.forEach((doc) => {
        //eslint-disable-next-line no-param-reassign
        delete doc._id;
        //eslint-disable-next-line no-param-reassign
        delete doc.semanticSearch._id;
        //eslint-disable-next-line no-param-reassign
        delete doc.semanticSearch.searchId;
      });
      delete res._id;
      expect(res.results.length).toBe(2);
      expect(res).toMatchSnapshot();
    });
    it('should return 404 if search does not exist', async () => {
      try {
        await semanticSearch.getSearch(db.id());
        fail('should throw error');
      } catch (e) {
        expect(e).toEqual(createError('Search not found', 404));
      }
    });
  });
});

