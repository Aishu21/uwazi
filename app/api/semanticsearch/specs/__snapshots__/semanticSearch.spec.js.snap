// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`semanticSearch create should create and save new search 1`] = `
Object {
  "__v": 0,
  "creationDate": 1000,
  "documents": Array [
    Object {
      "sharedId": "doc1",
      "status": "pending",
    },
    Object {
      "sharedId": "doc2",
      "status": "pending",
    },
  ],
  "language": "en",
  "searchTerm": "Test term",
  "status": "pending",
}
`;

exports[`semanticSearch getSearch should fetch a search by id and its document entities with semantic search results 1`] = `
Object {
  "documents": Array [
    Object {
      "sharedId": "doc1",
      "status": "completed",
    },
    Object {
      "sharedId": "doc2",
      "status": "completed",
    },
  ],
  "language": "en",
  "results": Array [
    Object {
      "language": "en",
      "semanticSearch": Object {
        "results": Array [
          Object {
            "page": 1,
            "score": 0.5,
            "sentence": "stuff",
          },
          Object {
            "page": 2,
            "score": 0.2,
            "sentence": "this",
          },
        ],
        "sharedId": "doc1",
      },
      "sharedId": "doc1",
    },
    Object {
      "language": "en",
      "semanticSearch": Object {
        "results": Array [
          Object {
            "page": 1,
            "score": 0.8,
            "sentence": "that",
          },
          Object {
            "page": 2,
            "score": 0.1,
            "sentence": "what",
          },
        ],
        "sharedId": "doc2",
      },
      "sharedId": "doc2",
    },
  ],
  "searchTerm": "torture",
  "status": "completed",
}
`;

exports[`semanticSearch processDocument should save the results of the document sorted by score in descending order, and compute average score 1`] = `
Array [
  Object {
    "page": 2,
    "score": 0.6,
    "text": "page 2",
  },
  Object {
    "page": 1,
    "score": 0.2,
    "text": "page 1",
  },
]
`;

exports[`semanticSearch processSearchLimit should return the updated search and processed documents 1`] = `
Object {
  "processedDocuments": Array [
    "doc2",
    "doc3",
  ],
  "updatedSearch": Object {
    "_id": "5c8162522b4081771f080487",
    "documents": Array [
      Object {
        "sharedId": "doc1",
        "status": "completed",
      },
      Object {
        "sharedId": "doc2",
        "status": "completed",
      },
      Object {
        "sharedId": "doc3",
        "status": "completed",
      },
      Object {
        "sharedId": "doc4",
        "status": "completed",
      },
      Object {
        "sharedId": "doc5",
        "status": "pending",
      },
    ],
    "language": "en",
    "searchTerm": "injustice",
    "status": "inProgress",
  },
}
`;
